packages/nimble-components/src/table/specs/table-columns-hld.md

```diff
@@ -0,0 +1,256 @@
+# Nimble Table Columns API
+
+## Problem Statement
+
+The `nimble-table` requires users to be able to configure which columns to display for a table. Additionally, a particular display column may require access to multiple fields in the data, and allow a user to either define the field to sort by, or provide a custom sort routine. We need to provide a means for a client to provide their own column visualization. Finally, columns must also provide the user to either specify the header text, or custom visuals to use for the header.
+
+### Out of scope of this HLD
+
+Programmatic API for state that could be considered column-centric: width, sort direction, grouped, etc. These concerns should be covered in separate designs covering those topics specifically, allowing for discussion on both the interactive side and the API design on an individual basis.
+
+## Links To Relevant Work Items and Reference Material
+
+[Table Declarative Columns API prototype:](https://github.com/ni/nimble/blob/325983040e886e52a100664d8fb1129dee767c2f/packages/nimble-components/src/table/tests/table.stories.ts#L23) ([Storybook](https://60e89457a987cf003efc0a5b-twewjutggo.chromatic.com/iframe.html?args=&id=table--table-story&viewMode=story))
+
+## Implementation / Design
+
+Columns will be provided to the table as slotted elements. The slot for the columns will be the default slot for the table, and thus no slot needs to be expressly provided by the user for a column:
```

Comment:
Is your vision that the order of the columns in the document will determine the order in the table? Any complications when we support interactive/programmatic column re-ordering or would we just re-order the columns in the document to achieve that?

Reply:
At the moment, I think we expect clients to re-order the columns in the document as needed. It's understood that this might make the template code considerably more noisy/complex than it otherwise would be.

Reply:
Makes sense as a starting point. Let's mention it in the spec.